### DISCUSSION TITLE
RLS Policy for Auth0 Authenticated Users

### DISCUSSION THREAD

**Discussion by SevanBadal** [supabase/supabase#6441] [Questions] [answered]
Users of my application (NextJS) authenticate using Auth0.  I need to enable inserts of certain resources only for authenticated users.  By following the [Supabase Auth0 Integration guide](https://supabase.com/docs/guides/integrations/auth0), I can successfully restrict reads to certain users by inspecting `user_id` on the JWT payload.  

However, the default policy for restricting inserts to authenticated users  `(role() = 'authenticated'::text)` does not work.  I am getting a `401` from my client-side insert operations even when a User is logged in.

Any help on this issue would be greatly appreciated!

**Comment by GaryAustin1** [comment, marked as answer]
Can you use whatever method you are using to "inspect" user_id in the JWT to just verify it exists as that would mean authenticated?

**Reply by SevanBadal** [reply]
A policy with the following expression is possible `(user_id() IS NOT NULL)`. I guess using `(role() = 'authenticated'::text)` felt safer. Is it not possible to specify a role when setting the access token on the client side?

**Reply by GaryAustin1** [reply]
Maybe I'm missing something in your process. 
Normal Auth with gotrue fills in the auth.user table with an id,role(always authenticated) and other user info when you signup a user.
When you signin a user from the client, you get a jwt that has the users id, role, etc.  
When you do a request to the database the jwt is provided which contains the role.  
The user can not change their role, if they could then he could change their id.

I did a quick glance at the auth0 tutorial and you could probably add a role when you create the jwt and then write another function to get that role (or possibly use the built in one if you use same naming/style for it).  Then when you create the jwt it would have a role in it.  
BUT basically unless you are going to do different roles, if you have a user_id in the token, you are authenticated.
EDIT also you said when you create the access token client side, I hope that part is running on the server side code as the true client code in the browser should not have access to that code.

**Reply by SevanBadal** [reply]
The token is created on the server side and then passed to the supabase client on the client side. Thanks for your help @GaryAustin1
